Sources/TranscribeCLI/main.swift

```diff
@@ -3,42 +3,56 @@
 ///
 ///     The resulting Markdown and WebVTT file is saved to the current directory with the
 ///     same basename as the input file.
 
 import Foundation
 import Transcribe
-
-struct TranscribeCLIError: Error {
-    var code: Int
-    var message: String
-
-    static let missingArgument = TranscribeCLIError(code: 1, message: "Usage: \(CommandLine.arguments[0]) FILE")
-    static func argumentIsNotAFile(argument: String) -> TranscribeCLIError {
-        return TranscribeCLIError(code: 2, message: "\(argument) must be a file")
-    }
-}
+import Utility
 
 // MARK: - Main program
 
 do {
-    guard CommandLine.arguments.count == 2 else {
-        throw TranscribeCLIError.missingArgument
+    // setup the argument parser
+    let parser = ArgumentParser(usage: "<options>",
+                                overview: "A Swift parser for output files from automated transcription services.")
+    let rawInputFilename = parser.add(option: "--json", shortName: "-j", kind: PathArgument.self, usage: "Amazon transcribe json file, e.g. './TestFixtures/amazon-transcribe-swift-community-podcast-0001-formatted-short.json'")
+    let rawSpeakerNames = parser.add(option: "--names", shortName: "-n", kind: [String].self, usage: "Array of speaker names, e.g. 'alice bob'")
+    
+    // The first argument is always the executable, drop it
+    let arguments = Array(ProcessInfo.processInfo.arguments.dropFirst())
+    let parsedArguments = try parser.parse(arguments)
+    guard let inputFilename = parsedArguments.get(rawInputFilename),
+        inputFilename.path.asString.lowercased().hasSuffix("json") else {
+        throw ArgumentParserError.invalidValue(argument: "--json", error: ArgumentConversionError.custom("No json file specified"))
     }
-
-    let inputFilename = CommandLine.arguments[1]
-    let inputFile = URL(fileURLWithPath: inputFilename).standardizedFileURL
-    guard let isRegularFile = try inputFile.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile, isRegularFile == true else {
-        throw TranscribeCLIError.argumentIsNotAFile(argument: inputFile.path)
+   
+    // validate json filename
+    let inputFile = URL(fileURLWithPath: inputFilename.path.asString).standardizedFileURL
+    guard let isRegularFile = try? inputFile.resourceValues(forKeys: [.isRegularFileKey]).isRegularFile, isRegularFile == true else {
+        throw ArgumentParserError.invalidValue(argument: "--json", error: ArgumentConversionError.custom("File not found"))
     }
 
-    let transcript = try AmazonTranscribe.Transcript(file: inputFile)
+    // parse the amazon transcription json file
+    var transcript = try AmazonTranscribe.Transcript(file: inputFile)
     let outputDirectory = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)
 
+    // Set speaker names
+    if let speakerNames = parsedArguments.get(rawSpeakerNames),
+        speakerNames.count == transcript.speakers.count {
+        for (index, speaker) in transcript.speakers.enumerated() {
```

Comment:
This is a pet peeve of mine: don't use "index" and "enumerated" together in the same expression. Although this code works for arrays, it wouldn't work for other collections where indices aren't integers (like `String`) or where indices don't start at 0 (like `ArraySlice`). `enumerated` gives you an _offset_, not an index. If you want an index (like here), write:

```swift
for (index, speaker) in zip(transcript.speakers.indices, transcript.speakers) {
    ...
```

Reply:
Edit: This whole expression feels wrong to me because we're using an index into `transcript.speaker` to index into another array. It works at the moment but that could change in the future. I'd try to rewrite this.

Reply:
What do you think about this solution?

```swift
    // Set speaker names
    if let speakerNames = parsedArguments.get(rawSpeakerNames) {
        
        // counts of cli names and transcript names do not match => print warning
        if speakerNames.count != transcript.speakers.count {
            print("Warning: count of input names (\(speakerNames)) does not match the speaker names in transcript: \(transcript.speakers.count)")
        }
        
        var iterator = speakerNames.makeIterator()
        for speaker in transcript.speakers {
            guard let speakerName = iterator.next() else { break }
            transcript[speaker: speaker.speakerLabel]?.name = speakerName
        }
    }
```

Reply:
The warning message should be printed to `stdError` (like the error message below). I think I have an even cleaner solution. @JulianKahnert Can you tick the "Allow edits from maintainers" box so that I can push to this branch?

<img width="257" alt="screen shot 2019-02-13 at 20 29 09" src="https://user-images.githubusercontent.com/5010/52738090-29a5a200-2fce-11e9-807d-914fc6717149.png">

Reply:
Actually this one is checked...hm

Reply:
Okay. I have no permission to push (maybe because you used the master branch on your side? Not sure).

Anyway, here's my solution (based on yours):

```swift
   if let speakerNames = parsedArguments.get(rawSpeakerNames) {
        // Counts of names passed on CLI and speakers in Transcript do not match => print warning
        if speakerNames.count != transcript.speakers.count {
            print("Warning: Number of speaker names passed in (\(speakerNames.count)) does not match the number of speakers in the transcript (\(transcript.speakers.count) – \(transcript.speakers.map { $0.speakerLabel }))", to: &stdError)
        }

        for (speakerName, speakerIndex) in zip(speakerNames, transcript.speakers.indices) {
            transcript.speakers[speakerIndex].name = speakerName
        }
    }
```

Can you turn this into a commit, please?

Reply:
done in f267769b3cb24f9005c1bbac38af956d5d038ed4